### ISSUE TITLE
[RFC] Forms and Service layer

### ISSUE THREAD

**Issue by adamlundrigan** [ZF-Commons/ZfcUser#561]
I've combined proposed changes to the forms and service layer into one as the latter depends heavily on the former.  This one actually is an RFC as I haven't implemented a prototype yet.

# Service Layer

We can simplify the service layer by pushing concerns such as the object being hydrated/extracted and password hashing on save -- both of which the service currently handles -- out of the service layer.
 - CreateUserForm can instantiate an empty instance of the user class itself
 - Password hashing can be performed by the hydrator when required

```
interface UserServiceInterface
{
    /**
     * @return UserEntityInterface
     * @throws RuntimeException on failure to create user
     */
    public function create(array $data);
    
    /**
     * @return UserEntityInterface
     * @throws RuntimeException on failure to update user
     */
    public function update(UserEntityInterface $user, array $data);
    
    /**
     * @return UserEntityInterface
     * @throws RuntimeException on failure to delete user
     */
    public function delete(UserEntityInterface $user);
    
    // Convenience; proxy directly to mapper methods
    public function findById($value);
    public function findByUsername($value);
    public function findByEmail($value);
}
```

> Q: Proxy `find*` methods through to mapper, or just provide a getter for the mapper as the current service does?  My preference is on proxying the methods so we're not forcing userland code to reach through the service to get at the mapper to run a query. 

A class implementing this interface would require three dependencies
 - Mapper instance
 - CreateUserForm instance
 - UpdateUserForm instance

> NOTE: CreateUserForm and UpdateUserForm are instances of the same class (UserForm).  They are created separately by the FormElementManager so that modules extending ZfcUser could attach custom fields differentially based on the action occurring. 

# Form Layer

The login form and input filter would remain largely unchanged except for some updates now that we're away from PHP 5.3.x.  The registration form and input filter will evolve to work for both create and update operations: 

## UserForm (formerly BaseForm)
In terms of elements it will be largely unchanged.
 - Change `password` to `new_password` and `passwordVerify` to `new_password_verify`

## UserInputFilter (formerly RegisterFilter)
In terms of elements it will be largely unchanged.
 - Change `password` to `new_password` and `passwordVerify` to `new_password_verify`

The `RecordExists` and `NoRecordExists` validators should be updated to use the context argument so that they work for both create and update operations. 

> Q: Do we need both? RecordExists isn't actually used by ZfcUser. 

> Q: Should the validators be changed to allow calling arbitrary mapper methods, like the new authentication layer does? 

----
#Notes
## Why the change in password fields?
To separate the action of changing the user's password from it's storage in the entity.  The password hash isn't something you want displayed in the HTML form, so it doesn't belong in the form. 

In the above scenario the form's InputFilter would validate `new_password` and `new_password_verify` together and throw back an error if they're non-empty and don't match.  If that validation passes the hydrator would take the value of `new_password`, hash it and store it in the entity's `password` field.  If `new_password` is empty the `password` field remains unchanged.

**Comment by claytondaley** [comment]
Zend/Form already has intelligence to handle passwords so this concern is resolve by PR #563, possibly eliminating the need to make yet another BC change.

**Comment by adamlundrigan** [comment]
@claytondaley good point, the `new_password` stuff isn't really necessary then

**Comment by claytondaley** [comment]
I also really like the idea of implementing password hashing through a Hydrator Strategy.  In fact, it would be great if the Hydrator was implemented to be easily extensible.  

 - For example, I have [PR](https://github.com/Danielss89/ZfcUserAdmin/pull/66) that converts ZfcUserAdmin's EditForm to use a hydrator.
 - Now I need to extend it to [properly Hydrate DateTimes](https://juriansluiman.nl/article/125/strategies-for-hydrators-a-practical-use-case).

It'd really be better if ZfcUserAdmin used (extended) a core hydrator... and I could extend it further. Unfortunately, I'm still on the learning curve for Hydrator Strategies so I'd benefit from a recommended pattern for this. My go-to was the event system, but my research on performance suggests this may not be the best long-term strategy.

**Comment by adamlundrigan** [comment]
> I also really like the idea of implementing password hashing through a Hydrator Strategy. In fact, it would be great if the Hydrator was implemented to be easily extensible.

I've been battling with that myself, as it's not the purpose of the hydrator.  The hydrator is meant to have one responsibility: converting between an object and it's array representation.  The proper solution would be to do this on the form post-validation but pre-object-populate, so I'm thinking the best course of action would be to create a filter which hashes the value of the password field. 

> For example, I have PR that converts ZfcUserAdmin's EditForm to use a hydrator.  Now I need to extend it to properly Hydrate DateTimes.

I had worked on something similar last summer but never got the chance to complete it and PR it (I did recently get as far as opening a ticket for it: https://github.com/Danielss89/ZfcUserAdmin/issues/65).  A lot of what ZfcUserAdmin does should be part of ZfcUser, and that module should just be the UI and integration with ZfcAdmin; the logic of CRUDing users is something ZfcUser should handle itself...and that's the main goal of my recent PR blitz here, which when complete I had planned to turn my attention to ZfcUserAdmin next.

**Comment by claytondaley** [comment]
You're right.  This is really what InputFilters do.  The [Zend\Form Quick Start](http://framework.zend.com/manual/current/en/modules/zend.form.quick-start.html) gives examples like:

```
             'filters' => array(
                 array('name' => 'Zend\Filter\StringTrim'),
                 array('name' => 'Zend\Filter\StringToLower'),
             ),
```

If an InputFilter can trim and change case, why not hash.  If the stored hash uses a different strategy, you can always `getRawValues()` and rehash.

**Comment by claytondaley** [comment]
... and I finally figured out the logic way to handle hydrator extension.

 - In the constructor, 
     - Use `setHydrator()` to put the desired hydrator in place
     - Apply any desired hydrator strategies
     - Kick off an `init` event to welcome extensions of the form
 - In the extending module,
     - Attach a listener to the form `init`
     - In callback, add elements
     - In the same callback (right after you add the element), use `$form->getHydrator()` to access the hydrator and add the strategy.

Put together, this looks something like:

    public function extendEditForm($event) {
        $form = $event->getTarget();
        ...
        $form->add(
            array(
                'type' => 'date',
                'name' => 'start_date',
                'options' => array(
                    'label' => 'Start Date',
                ),
            )
        );
        $form->getHydrator()->addStrategy('last_active', new DateTimeStrategy);
        ...
    }

**Comment by claytondaley** [comment]
For the ZfcUser forms, I'd also like to recommend a dynamic validation group implementation comparable to [Rob Allen's recommendation](http://akrabat.com/zend-framework-2/exclude-elements-from-zendforms-getdata/).  I care more about the requirement than the strategy -- add/remove methods for the validation group would be just as good (and perhaps better).

This is particularly important on "edit" forms where you may want to present `readonly` data that should not be committed back to the database.  The current method for achieving this is to create a validation group.  Unfortunately, the default validation group implementation depends on explicit knowledge of all of the elements on the form, which is antithetical to easy extension.

I believe the `readonly` attribute should also exclude by default, with the option to override by setting `exclude=> false`.  IMHO the core Zend/Form should be upgraded in this way, but this is as good a place to start as any.

**Comment by claytondaley** [comment]
A final enhancement would be a universal form renderer place in a parent project (ZfcMvc?) that is inherited by ZfcUser and ZfcAdmin.  ZfcUserAdmin has the right idea importing `_form.phtml` into numerous page.  I believe I've successfully implemented a [fully generic example](https://github.com/claytondaley/ZfcUserAdmin/blob/e33578d89b4900d0b4d34405bad7dadf004c42a7/view/zfc-user-admin/user-admin/_form.phtml) that would be a full drop-in replacement for existing ZfcUser forms.  Several notes:

 - This strategy still hard-codes a DOM structure (in this case dl and dt). I assume it's OK (even desirable) in Zfc* to hard code a standard so it's easier to write a universal CSS skin. As part of a BC break, however, we might consider an upgrade ([ZfcUserAdmin#63](https://github.com/Danielss89/ZfcUserAdmin/pull/63) offers an alternative).
 - At the same time, I dislike the __way__ the DOM structure is hard-coded.  For example, nested collections won't benefit from the current approach.  It would be nice to figure out how to do this at more of a configurations level so that we need only call `$form->form()`.
 - I also think we should inject the redirect element into the form in the controller (hidden type, excluded from validation) then let the rendering engine actual add it.

**Comment by adamlundrigan** [comment]
RE: validation groups this is something I've brought up in discussion about ZF3's forms, as the current implementation of validation groups makes it hard to work with complex forms.  IMO there should be a convenience wrapper around that so you can selectively include and exclude fields without having to specify the entire validation group.  In one of my projects last year I wrote an atrociously dirty hack to work around the issue, and it's on my someday agenda to extract and clean up that code so it can be released. 

RE: form rendering, I think that's best left up to the consuming application.  We'll bundle a minimum viable set of view scripts (like we do now) but if those don't fit your use case you'll need to override them.  The approach ZfcUser currently uses (looping over every element in the form) is likely the behavior that will remain, though more than likely anyone adding custom fields would also be overriding the view scripts to control how they are displayed.  ZfcUserAdmin's handling of this annoys me a bit as it `include`s the common form template rather than sending it through the view renderer, making it unnecessarily difficult to override.

RE: redirects, I'm not sure about this one.  IMO there are better ways to handle the post-login redirect than sticking the target in a hidden form element.  That's prone to abuse if not done properly, as we saw last year with the open redirect security issue.  It's something I plan to give some thought to once the current round of PRs I've opened are wrapped up.

**Comment by claytondaley** [comment]
Obviously, there's no code to react to so the current state is the available point of reference

 - Redirect is important for an authentication module.
     - The current approach would be significantly improved by adding the element to the form around the time it's created.  It would automatically populate when `setData()` is called and be automatically returned with an invalid form.  There seems no advantage to the current approach.
     - I'm also skeptical that you can do much more than plug bugs in redirect logic.  A user provides the first URL by typing it into the browser and you make sure -- at the end of the process using a 302 -- that the user's browser is the one that calls that URL again.  I don't see how a modified form submission can be any worse than the original URL -- both provided by the user.

----
There are several points in your response on forms.  Trying to separate them out:

 - I believe the `register.phtml` loop would work perfectly for `login.phtml` as well.  Thus, ZfcUser 2.x has two blocks of code when one would suffice.  The ZfcUserAdmin strategy is DRY by sharing that code between both pages.
 - Granted, the ZfcUserAdmin strategy would be *even better* if it you could configure that script from the Controller... but the current state is no worse than ZfcUser's equally hard-coded current state.
 - Finally, the generic loop I developed takes the loop-replaces-login idea to the logical conclusion.  Minus a (now-fixed) bug for button labels, it should exactly reproduce ZfcUser and ZfcUserAdmin forms 
     - To make sure I didn't stick my foot in my mouth, I went ahead and replaced every 1.x form with the generic renderer (I'm using ZfcUserAdmin so I have to stay on 1.x).  There are some quirks... for example, the ChangeEmail form doesn't `add()` a submit button in `__construct`.  Besides that, the forms are as expected -- materially indistinguishable.

**Comment by claytondaley** [comment]
Ran into another quirk worth adding to the punchlist.  Custom Form Elements only work when you use [two specific strategies](http://framework.zend.com/manual/current/en/modules/zend.form.advanced-use-of-forms.html#creating-custom-elements):

 - The form must be created using the `FormElementManager` (e.g. `$sl->get('FormElementManager')->get('\Application\Form\MyForm');`)
 - The elements must be added during `init()` rather than `__construct()`.

If you want to add fields after the `FormElementManager` delivers the form, I found it also necessary to manually inject the `FormElementManager` to the Form.  This seems to be a problem because the `getFormFactory` and `getFormElementManager` create new instances using the FQN (analysis in several replies starting [here](http://zend-framework-community.634137.n4.nabble.com/Adding-custom-form-elements-tp4659495p4661597.html)).  

It would be helpful to users to include a workaround when we construct forms, the following loosely based on the ZfcUserAdmin module:

```
        // Must use the FormElementManager to support custom FormElements
        $formElementManager = $this->getServiceLocator()->get('FormElementManager');

        // Form building code
        $form = $formElementManager->get('\ZfcUserAdmin\Form\FormName');
        // Strangely, the FormElementManager isn't automatically injected into the Table
        $form->getFormFactory()->setFormElementManager($formElementManager);
        $form->addUserFields($this->getOptions()->getUserListElements());
```
